packages/backend/server/src/plugins/copilot/workflow/node.ts

```diff
@@ -0,0 +1,146 @@
+import path, { dirname } from 'node:path';
+import { fileURLToPath } from 'node:url';
+
+import { Logger } from '@nestjs/common';
+import Piscina from 'piscina';
+
+import { CopilotChatOptions } from '../types';
+import { getWorkflowExecutor, WorkflowExecutor } from './executor';
+import {
+  NodeData,
+  WorkflowNodeState,
+  WorkflowNodeType,
+  WorkflowResult,
+  WorkflowResultType,
+} from './types';
+
+export class WorkflowNode {
+  private readonly logger = new Logger(WorkflowNode.name);
+  private readonly edges: WorkflowNode[] = [];
+  private readonly parents: WorkflowNode[] = [];
+  private readonly executor: WorkflowExecutor | null = null;
+  private readonly worker: Piscina;
+
+  constructor(private readonly data: NodeData) {
+    if (data.nodeType === WorkflowNodeType.Basic) {
+      this.executor = getWorkflowExecutor(data.type);
+    }
+    this.worker = new Piscina({
+      filename: path.resolve(
+        dirname(fileURLToPath(import.meta.url)),
+        'worker.mjs'
+      ),
+      minThreads: 2,
+      // empty envs from parent process
+      env: {},
+      argv: [],
+      execArgv: [],
+    });
+  }
+
+  get id(): string {
+    return this.data.id;
+  }
+
+  get name(): string {
+    return this.data.name;
+  }
+
+  get config(): NodeData {
+    return Object.assign({}, this.data);
+  }
+
+  get parent(): WorkflowNode[] {
+    return this.parents;
+  }
+
+  // if is the end of the workflow, pass through the content to stream response
+  get hasEdges(): boolean {
+    return !!this.edges.length;
+  }
+
+  private set parent(node: WorkflowNode) {
+    if (!this.parents.includes(node)) {
+      this.parents.push(node);
+    }
+  }
+
+  addEdge(node: WorkflowNode): number {
+    if (this.data.nodeType === WorkflowNodeType.Basic) {
+      if (this.edges.length > 0) {
+        throw new Error(`Basic block can only have one edge`);
+      }
+    } else if (
+      this.data.nodeType === WorkflowNodeType.Decision &&
+      !this.data.condition
+    ) {
+      throw new Error(`Decision block must have a condition`);
+    }
+    node.parent = this;
+    this.edges.push(node);
+    return this.edges.length;
+  }
+
+  private async evaluateCondition(
+    condition: string,
+    params: WorkflowNodeState
+  ): Promise<string | undefined> {
+    try {
+      const chooseNode = this.worker.run({
+        iife: `(${condition})(nodeIds, params)`,
```

Comment:
why do we need to send condition to a worker to execute? the conditions are programmatically executable and predefined, all args are known at the mean time

Reply:
we will allow users to modify prompts in the admin dashboard in future, then the workflow graph should also allow console modifications
in this case, the condition will not only be predefined, but may also be unknown risk inputs
so we need to use workers to isolate untrusted code

Reply:
It must be a long story allowing users modifications, and I prefer to use worker when it's really needed

Reply:
switched to evaluate via new Function in unmodified state